### PR TITLE
Fix lossy type promotion in JAX primitive lowering

### PR DESCRIPTION
**Pull request by dime10**
https://github.com/PennyLaneAI/pennylane/pull/6082 unearthed that in an attempt to be more lenient with user supplied types, Catalyst eagerly converts any type to the required float64 type for gate parameters, including when this results in a loss of data (like converting complex numbers to floats).

This fixes the issue as well as providing some minor code cleanup, and fixing a long-standing issue of potentially undefined variables in the Python code.

In order to help users with the proposed fix in [#pennylane/6082 ](https://github.com/PennyLaneAI/pennylane/pull/6082) for the decomposition of `Exp`, the error message for complex gate parameters mentions potential non-unitary operators like the exponential with real exponent.

[sc-71066]

### CODE REVIEW

**Review by erick-xanadu — APPROVED** [review]
I think it is ok.

I wouldn't do it at this level of abstraction though. When tracing gphase and extract you can use `jax.numpy.promote_types` to find the least upper bound of the type promotion lattice using the element type from the user and the type you want as the two operands. If the least upper bound is different from the type you want to promote to, you know that an error will be raised. If the least upper bound is the same, then you can safely promote with `jax.lax.convert_element_type` or `jax.numpy.astype`. This would lead to the operands in gphase and extract to already be tensors of the type you want in their lowering.

**Comment by erick-xanadu** [comment]
I'll take a look first time tomorrow :thinking:

**Comment by dime10** [comment]
> I'll take a look first time tomorrow 🤔

Thanks Erick! I've tried debugging this for a while on the AWS machine, but I could not make sense of it. Something related to the IR contexts that I'm generating in the tests, but this seems to be the standard pattern in mlir (and there are instances in jax as well) for unit testing functions using the Python bindings. One avenue I explored was maybe it doesn't like having multiple contexts alive at the same time, but even force garbage collection in between each test didn't seem to work.

For now I've disabled the tests on linux, I don't think the functional code has any issues, only the tests themselves.
